ComputerVision.cs

```diff
@@ -0,0 +1,518 @@
+    IEnumerator Start()
+    {
+#if UNITY_EDITOR
+        GetEdgeConnection();
+        yield break;
+#endif
+        if (Application.internetReachability == NetworkReachability.NotReachable)
+        {
+            // Handle failure due no internet connection
+            throw new Exception("No Internet Connection");
+        }
+        else 
+        {
+            // MobiledgeX uses location for finding the closest cloudlet to the user
+            yield return StartCoroutine(MobiledgeX.LocationService.EnsureLocation());
+            GetEdgeConnection();
+        }
+    }
+
+    void OnGUI()
+    {
+        // show Detection Rects after the screen shot have been taken and the server have responded
+        if (showGUI)
+        {
+            DrawRectangles();
+        }
+    }
+
+    private void Update()
+    {
+        if (client == null)
+        {
+            return;
+        }
+        var cqueue = client.receiveQueue;
+        string msg;
+        while (cqueue.TryPeek(out msg))
+        {
+            cqueue.TryDequeue(out msg);
+            HandleServerRespone(msg);
+        }
+
+    }
+    #endregion
+
+    async Task GetEdgeConnection()
+    {
+        integration = new MobiledgeXIntegration();
+        integration.UseWifiOnly(true);
+        
+        try
+        {
+            bool cloudletFound = await integration.RegisterAndFindCloudlet("eu-mexdemo.dme.mobiledgex.net", 38001);
+            if (cloudletFound)
+            {
+                integration.GetAppPort(LProto.L_PROTO_HTTP);
+                SetConnection();
+                StartCoroutine(ImageSenderFlow());
+            }
+        }
+
+        catch (Exception e) // In case we don't support the detected carrierName  (In the generated dme)
+        {
+            Debug.LogError(e + "Error finding Connecting to Edge.");
+            throw e;
+        }
+
+    }
+
+    public string UriBasedOnConnectionMode()
+    {
+        AppPort appPort;
+        string url;
+        switch (connectionMode)
+        {
+            case ConnectionMode.WebSocket:
+                appPort = integration.GetAppPort(LProto.L_PROTO_TCP);
+                url = integration.GetUrl("wss");
+                return url;
+            case ConnectionMode.Rest:
+                appPort = integration.GetAppPort(LProto.L_PROTO_TCP);
+                url = integration.GetUrl("https");
+                return url;
+            default:
+                return "";
+        }
+    }
+
+    public async void StartWebSocket(string url)
+    {
+        Uri uri = new Uri(url);
+        client = new MobiledgeXWebSocketClient();
+        if (client.isOpen())
+        {
+
+            client.Dispose();
+            client = new MobiledgeXWebSocketClient();
+        }
+        await client.Connect(uri);
+        wsStarted = true;
+    }
+
+    public void SendtoServer(byte[] imgBinary)
+    {
+        client.Send(imgBinary);
+    }
+
+    public IEnumerator SendImageToServer(byte[] imgBinary, string url)
```

Comment:
We do have a GetHttpClient function. But if this is more dependable, seems fine

Reply:
Good Point, After long time debugging, I think the best way for WebRequests in Unity is Coroutines, it doesn't block the main thread and have better performance.